### PR TITLE
sw_engine: split long lines to avoid overflow

### PR DESCRIPTION
**Pull request by mgrudzinska**
@Issue: https://github.com/thorvg/thorvg/issues/2651

### CODE REVIEW

**Comment by mgrudzinska** [comment]
@hermet I tried to test performance, but the differences between runs are too big to find a difference between this pr and main. can you please double check?
also there is not so many changes - diff looks bad, but I didn't make changes in the inside logic of _lineTo, I just applied the approach that we already have in _cubicTo - if line to too big, split.

**Comment by hermet** [comment]
> the differences between runs are too big to find a difference between this pr and main.

@mgrudzinska Could you please suggest the Lottie FPS? The value represents the average based on accumulation. You will see a stable value once the sampling exceeds a certain basis (i.e 1000)

image

**Comment by mgrudzinska** [comment]
> > the differences between runs are too big to find a difference between this pr and main.
> 
> @mgrudzinska Could you please suggest the Lottie FPS? The value represents the average based on accumulation. You will see a stable value once the sampling exceeds a certain basis (i.e 1000)
> 
> image

I have 17/18 on main and on this branch as well. so I can not see any changes

File: src/renderer/sw_engine/tvgSwRle.cpp [hunk 1]
```diff
@@ -513,103 +514,116 @@ static void _lineTo(RleWorker& rw, const SwPoint& to)
...
+                do {
+                    f2.y = 0;
+                    rw.cover += (f2.y - f1.y);
+                    rw.area += (f2.y - f1.y) * f1.x * 2;
+                    f1.y = ONE_PIXEL;
+                    --e1.y;
+                    _setCell(rw, e1);
+                } while(e1.y != e2.y);
+            }
+            //any other line
```
Comment (hermet) [inline src/renderer/sw_engine/tvgSwRle.cpp:565]: indentation

File: src/renderer/sw_engine/tvgSwRle.cpp [hunk 2]
```diff
@@ -513,103 +514,116 @@ static void _lineTo(RleWorker& rw, const SwPoint& to)
...
+            if (diff.y > 0) {
+                do {
+                    f2.y = ONE_PIXEL;
+                    rw.cover += (f2.y - f1.y);
+                    rw.area += (f2.y - f1.y) * f1.x * 2;
+                    f1.y = 0;
+                    ++e1.y;
+                    _setCell(rw, e1);
+                } while(e1.y != e2.y);
+                //vertical line down
```
Comment (hermet) [inline src/renderer/sw_engine/tvgSwRle.cpp:554]: indentation

File: src/renderer/sw_engine/tvgSwRle.cpp [hunk 3]
```diff
@@ -513,103 +514,116 @@ static void _lineTo(RleWorker& rw, const SwPoint& to)
...
+        e1 = TRUNC(line[1]);
+        e2 = TRUNC(line[0]);
+
+        auto f1 = line[1] - SUBPIXELS(e1);
+        SwPoint f2;
+
+        //inside one cell
+        if (e1 == e2) {
+            ;
+            //any horizontal line
```
Comment (hermet) [inline src/renderer/sw_engine/tvgSwRle.cpp:539]: indentation

**Comment by hermet** [comment]
@tinyjin Could you please confirm that the crash is resolved in the ThorVG viewer with this PR?
sample: test

**Review by hermet — APPROVED** [review]

File: src/renderer/sw_engine/tvgSwRle.cpp (hunk 1 above)
Comment (mgrudzinska) [reply src/renderer/sw_engine/tvgSwRle.cpp]: fix this and all other

**Comment by tinyjin** [comment]
@hermet 
Yes, that sample has no crash

CleanShot 2024-10-10 at 14 20 26@2x

**Review by tinyjin — APPROVED** [review]

**Pull request merged**